### ISSUE TITLE
CA2000 does not handle "is not null"

### ISSUE THREAD

**Issue by MartyIX** [dotnet/roslyn-analyzers#4981]
### Analyzer

**Diagnostic ID**: [CA2000](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2000): `Dispose objects before losing scope`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.201](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

**Update: See repro below: https://github.com/dotnet/roslyn-analyzers/issues/4981#issuecomment-816461346**

CA2000 seems not to recognize `is not null` pattern but it recognize `!= null` in the following snippet:

```csharp
WebSocketClient? client = null; // `WebSocketClient` implements `IAsyncDisposable`

try
{
    client = new(uri, nameof(MyApiService));
    Result connectRes = await client.ConnectAsync(cancellationToken).ConfigureAwait(false);
    if (connectRes)
    {
        result = Result<IWebSocketClient>.Ok(client);
        client = null;
    }
    else result = connectRes;
}
finally
{
    if (client is not null) // <--- CHANGE THIS TO `if (client != null)` and the CA2000 warning is gone.
    {
        await client.DisposeAsync().ConfigureAwait(false);
    }
}
```

This is certainly not a shortest code one can come up but it is easy to understand I think.

### Steps To Reproduce

Please see the example above.

### Expected behavior

`if (client is not null)` does not lead to CA2000 being emitted.

### Actual behavior

`is not null` seems to be not supported but `!= null` is.

**Comment by MartyIX** [comment]
I have modified a test:

```csharp
        [Fact, WorkItem(2201, "https://github.com/dotnet/roslyn-analyzers/issues/2201")]
        public async Task NestedTryFinallyInTryCatch_NoDiagnostic()
        {
            await VerifyCS.VerifyAnalyzerAsync(@"
using System.IO;

class Test
{
    void M1()
    {
        try
        {
            var ms = new MemoryStream();
            try
            {
            }
            finally
            {
                if (ms is not null) {
                    ms.Dispose();
                }
            }
        }
        catch
        {
        }
    }
}");
```

but it returns

```
  Message: 
    Microsoft.CodeAnalysis.Testing.Verifiers.EqualWithMessageException : Context: Diagnostics of test state
    Mismatch between number of diagnostics returned, expected "0" actual "1"
    
    Diagnostics:
    // /0/Test0.cs(16,27): error CS8370: Feature 'not pattern' is not available in C# 7.3. Please use language version 9.0 or greater.
    DiagnosticResult.CompilerError("CS8370").WithSpan(16, 27, 16, 35).WithArguments("not pattern", "9.0"),
    
    
    Assert.Equal() Failure
    Expected: 0
    Actual:   1
```

So I can't probably fix it myself.

**Comment by Youssef1313** [comment]
@MartyIX In case you want to debug this. Your test should look like:

```csharp
        [Fact, WorkItem(2201, "https://github.com/dotnet/roslyn-analyzers/issues/2201")]
        public async Task NestedTryFinallyInTryCatch_NoDiagnostic()
        {
            var code = @"
using System.IO;

class Test
{
    void M1()
    {
        try
        {
            var ms = new MemoryStream();
            try
            {
            }
            finally
            {
                if (ms is not null) {
                    ms.Dispose();
                }
            }
        }
        catch
        {
        }
    }
}
";

    await new VerifyCS.Test
    {
        TestCode = code,
        FixedCode = code,
        LanguageVersion = Microsoft.CodeAnalysis.CSharp.LanguageVersion.CSharp9,
    }.RunAsync();

```

**Comment by mavasani** [comment]
@MartyIX Can you please also verify whether this repros with the latest NuGet package for `Microsoft.CodeAnalysis.NetAnalyzers`?

**Comment by MartyIX** [comment]
This is shorter repro that one can use. I can reproduce with `Microsoft.CodeAnalysis.NetAnalyzers` if I did that correctly:

`.editorconfig`
```
root = true

[*.cs]
# CA2000: Dispose objects before losing scope
# Because an exceptional event might occur that will prevent the finalizer of an object from running, the object should be explicitly disposed
# before all references to it are out of scope.
dotnet_diagnostic.CA2000.severity = warning
```

```csharp
#nullable enable

using System;
using System.Threading;
using System.Threading.Tasks;

namespace TestApp
{
    class Program
    {
        public static async Task Main(string[] args)
        {
            Console.WriteLine("Hello World!");

            Timer? timer = null; 

            try
            {
                timer = new(null!);

                bool result = await ComputeAsync().ConfigureAwait(false);
                if (result)
                {
                    timer = null;
                }
            }
            finally
            {
                if (timer is not null) // <--- CHANGE THIS TO `if (timer != null)` and the CA2000 warning is gone.
                {
                    await timer.DisposeAsync().ConfigureAwait(false);
                }
            }
        }

        private static Task<bool> ComputeAsync()
        {
            while (true)
            {

            }
        }
    }
}
```

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
    <EnableNETAnalyzers>false</EnableNETAnalyzers>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="5.0.3">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
  </ItemGroup>

</Project>
```

**Comment by manfred-brands** [comment]
@mavasani Version 6.0.0 of the analyzer still gives the same error, regardless of SDK (5.0.404 or 6.0.101)

```
using System.IO;

class Test
{
    void M1()
    {
        MemoryStream ms = null;
        try
        {
            ms = new MemoryStream();
        }
        finally
        {
            if (ms is not null) {
                ms.Dispose();
            }
        }
    }
}
```

Debugging this, it actually throws an exception in `DataFlowOperationVisitor.TryInferConversion`.
The method deals with `IConstantPatternOperation`, but doesn't know about `NegatedPatternOperation`.
Changing the test to `!(ms is null)` causes the test to pass.

I tried to change the code, but in the release/6.0.1xx the type `INegatedPatternOperation` is not available because the analyzers reference indirectly Microsoft.CodeAnalysis.Common version 3.3.1 and it needs a later version.

**Comment by Youssef1313** [comment]
#5816 can help much with that I think.

**Comment by Youssef1313** [comment]
@manfred-brands I think you can also add the operation kind in:

https://github.com/dotnet/roslyn-analyzers/blob/main/src/Utilities/Compiler/Lightup/OperationKindEx.cs

```csharp
        public const OperationKind NegatedPattern = (OperationKind)109;
```

**Comment by manfred-brands** [comment]
@Youssef1313 I resorted to adding a class in the Lightup directory instead as I needed access to its `Pattern` property.
